File: src/notify.py
# -*- coding: utf-8 -*-
"""
通知模块：支持飞书通知和自定义 Webhook 通知。
"""
import json
import requests
from typing import Optional, Dict, Any

from colorama import Fore, Style


# 飞书卡片颜色预设
FEISHU_COLORS = {
    "蓝色 (Blue)": "blue",
    "绿色 (Green)": "green",
    "红色 (Red)": "red",
    "橙色 (Orange)": "orange",
    "紫色 (Purple)": "purple",
    "靛蓝 (Indigo)": "indigo",
    "灰色 (Grey)": "grey",
}


def send_feishu_notification(
    webhook_url: str,
    title: str,
    content: str,
    color: str = "blue",
    footer: str = "来自小雪工具箱",
) -> bool:
    """
    发送飞书机器人卡片消息。

    Args:
        webhook_url: 飞书 Webhook URL
        title: 卡片标题
        content: 消息内容 (支持 lark_md 格式)
        color: 卡片头部颜色 (blue/green/red/orange/purple/indigo/grey)
        footer: 卡片底部备注

    Returns:
        bool: 发送是否成功
    """
    if not webhook_url:
        print(f"{Fore.YELLOW}[警告] 飞书 Webhook URL 为空，跳过发送{Style.RESET_ALL}")
        return False

    # 构建飞书卡片消息
    message = {
        "msg_type": "interactive",
        "card": {
            "config": {
                "wide_screen_mode": True
            },
            "header": {
                "title": {
                    "tag": "plain_text",
                    "content": title
                },
                "template": color
            },
            "elements": [
                {
                    "tag": "div",
                    "text": {
                        "tag": "lark_md",
                        "content": content
                    }
                },
                {
                    "tag": "hr"
                },
                {
                    "tag": "note",
                    "elements": [
                        {
                            "tag": "plain_text",
                            "content": footer
                        }
                    ]
                }
            ]
        }
    }

    try:
        response = requests.post(
            webhook_url,
            headers={"Content-Type": "application/json"},
            data=json.dumps(message),
            timeout=30
        )

        if response.status_code == 200:
            result = response.json()
            if result.get("code") == 0 or result.get("StatusCode") == 0:
                print(f"{Fore.GREEN}[成功] 飞书通知发送成功{Style.RESET_ALL}")
                return True
            else:
                print(f"{Fore.RED}[失败] 飞书返回错误: {result}{Style.RESET_ALL}")
                return False
        else:
            print(f"{Fore.RED}[失败] HTTP 状态码: {response.status_code} - {response.text}{Style.RESET_ALL}")
            return False

    except requests.exceptions.Timeout:
        print(f"{Fore.RED}[错误] 请求超时{Style.RESET_ALL}")
        return False
    except requests.exceptions.RequestException as e:
        print(f"{Fore.RED}[错误] 网络请求失败: {str(e)}{Style.RESET_ALL}")
        return False
    except Exception as e:
        print(f"{Fore.RED}[错误] 发送飞书通知时发生异常: {str(e)}{Style.RESET_ALL}")
        return False


def send_webhook_notification(
    url: str,
    headers: Optional[Dict[str, str]] = None,
    body: Optional[Dict[str, Any]] = None,
    headers_json: Optional[str] = None,
    body_json: Optional[str] = None,
) -> bool:
    """
    发送自定义 Webhook POST 请求。

    Args:
        url: POST 请求 URL
        headers: 请求头字典 (优先使用)
        body: 请求体字典 (优先使用)
        headers_json: 请求头 JSON 字符串 (备选)
        body_json: 请求体 JSON 字符串 (备选)

    Returns:
        bool: 发送是否成功
    """
    if not url:
        print(f"{Fore.YELLOW}[警告] Webhook URL 为空，跳过发送{Style.RESET_ALL}")
        return False

    # 解析 Headers
    final_headers = {"Content-Type": "application/json"}
    if headers:
        final_headers.update(headers)
    elif headers_json:
        try:
            parsed_headers = json.loads(headers_json)
            if isinstance(parsed_headers, dict):
                final_headers.update(parsed_headers)
        except json.JSONDecodeError as e:
            print(f"{Fore.RED}[错误] Headers JSON 解析失败: {str(e)}{Style.RESET_ALL}")
            return False

    # 解析 Body
    final_body = body
    if final_body is None and body_json:
        try:
            final_body = json.loads(body_json)
        except json.JSONDecodeError as e:
            print(f"{Fore.RED}[错误] Body JSON 解析失败: {str(e)}{Style.RESET_ALL}")
            return False

    if final_body is None:
        final_body = {}

    try:
        print(f"{Fore.CYAN}[请求] POST {url}{Style.RESET_ALL}")
        print(f"  Headers: {json.dumps(final_headers, ensure_ascii=False)}")
        print(f"  Body: {json.dumps(final_body, ensure_ascii=False)}")

        response = requests.post(
            url,
            headers=final_headers,
            json=final_body,
            timeout=30
        )

        print(f"\n{Fore.CYAN}[响应] 状态码: {response.status_code}{Style.RESET_ALL}")
        
        # 尝试解析 JSON 响应
        try:
            resp_json = response.json()
            print(f"  响应内容: {json.dumps(resp_json, ensure_ascii=False, indent=2)}")
        except:
            print(f"  响应内容: {response.text[:500]}")

        if 200 <= response.status_code < 300:
            print(f"\n{Fore.GREEN}[成功] Webhook 请求发送成功{Style.RESET_ALL}")
            return True
        else:
            print(f"\n{Fore.YELLOW}[警告] 服务器返回非 2xx 状态码{Style.RESET_ALL}")
            return False

    except requests.exceptions.Timeout:
        print(f"{Fore.RED}[错误] 请求超时{Style.RESET_ALL}")
        return False
    except requests.exceptions.RequestException as e:
        print(f"{Fore.RED}[错误] 网络请求失败: {str(e)}{Style.RESET_ALL}")
        return False
    except Exception as e:
        print(f"{Fore.RED}[错误] 发送 Webhook 时发生异常: {str(e)}{Style.RESET_ALL}")
        return False

File: engine/src/notify.rs
//! 任务完成通知 (移植自 src/notify.py)。
//!
//! 纯 payload 构建可测; 发送用 ureq (rustls), 30s 超时。

use serde_json::{json, Value};

/// 飞书卡片颜色 (notify.FEISHU_COLORS 常用子集)。
pub const FEISHU_COLORS: &[(&str, &str)] = &[
    ("蓝色 (Blue)", "blue"),
    ("绿色 (Green)", "green"),
    ("红色 (Red)", "red"),
    ("黄色 (Yellow)", "yellow"),
    ("橙色 (Orange)", "orange"),
    ("紫色 (Purple)", "purple"),
    ("灰色 (Grey)", "grey"),
    ("靛蓝 (Indigo)", "indigo"),
];

/// 构建飞书卡片消息 (notify.send_feishu_notification 的 payload 部分)。
pub fn feishu_payload(title: &str, content: &str, color: &str, footer: &str) -> Value {
    json!({
        "msg_type": "interactive",
        "card": {
            "config": {"wide_screen_mode": true},
            "header": {
                "title": {"tag": "plain_text", "content": title},
                "template": color
            },
            "elements": [
                {"tag": "div", "text": {"tag": "lark_md", "content": content}},
                {"tag": "hr"},
                {"tag": "note", "elements": [
                    {"tag": "plain_text", "content": footer}
                ]}
            ]
        }
    })
}

/// 发送飞书通知 (notify.send_feishu_notification)。
///
/// Ok(()) 发送成功; Err(原因) 失败 (URL 为空 / 网络 / 业务码非 0)。
pub fn send_feishu(webhook_url: &str, title: &str, content: &str,
                   color: &str, footer: &str) -> Result<(), String> {
    if webhook_url.is_empty() {
        return Err("飞书 Webhook URL 为空，跳过发送".into());
    }
    let payload = feishu_payload(title, content, color, footer);
    let resp = agent()
        .post(webhook_url)
        .send_json(payload)
        .map_err(|e| format!("网络请求失败: {e}"))?;
    if resp.status() != 200 {
        return Err(format!("HTTP 状态码: {}", resp.status()));
    }
    let result: Value = resp
        .into_json()
        .map_err(|e| format!("响应解析失败: {e}"))?;
    if result["code"] == 0 || result["StatusCode"] == 0 {
        Ok(())
    } else {
        Err(format!("飞书返回错误: {result}"))
    }
}

/// 发送自定义 Webhook POST (notify.send_webhook_notification)。
pub fn send_webhook(url: &str, headers: &[(String, String)],
                    body: &Value) -> Result<(), String> {
    if url.is_empty() {
        return Err("Webhook URL 为空，跳过发送".into());
    }
    let mut req = agent().post(url);
    let mut has_content_type = false;
    for (k, v) in headers {
        if k.eq_ignore_ascii_case("content-type") {
            has_content_type = true;
        }
        req = req.set(k, v);
    }
    if !has_content_type {
        req = req.set("Content-Type", "application/json");
    }
    let resp = req
        .send_json(body.clone())
        .map_err(|e| format!("网络请求失败: {e}"))?;
    let status = resp.status();
    if (200..300).contains(&status) {
        Ok(())
    } else {
        Err(format!("服务器返回非 2xx 状态码: {status}"))
    }
}

fn agent() -> ureq::Agent {
    ureq::AgentBuilder::new()
        .timeout_read(std::time::Duration::from_secs(30))
        .timeout_write(std::time::Duration::from_secs(30))
        .build()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn feishu_payload_shape() {
        let p = feishu_payload("完成", "任务 {task} 已完成", "blue", "小雪工具箱");
        assert_eq!(p["msg_type"], "interactive");
        assert_eq!(p["card"]["header"]["title"]["content"], "完成");
        assert_eq!(p["card"]["header"]["template"], "blue");
        assert_eq!(
            p["card"]["elements"][0]["text"]["content"],
            "任务 {task} 已完成"
        );
        assert_eq!(p["card"]["elements"][2]["elements"][0]["content"],
                   "小雪工具箱");
    }

    #[test]
    fn feishu_empty_url_rejected() {
        assert!(send_feishu("", "t", "c", "blue", "f").is_err());
    }

    #[test]
    fn webhook_empty_url_rejected() {
        assert!(send_webhook("", &[], &json!({})).is_err());
    }

    #[test]
    fn webhook_unreachable_reports_error() {
        // 本机保留端口, 连接必失败 (不依赖外网)
        let err = send_webhook(
            "http://127.0.0.1:1/hook",
            &[],
            &json!({"a": 1}),
        );
        assert!(err.is_err());
    }

    #[test]
    fn feishu_colors_cover_common() {
        assert!(FEISHU_COLORS.iter().any(|(n, c)| *c == "blue" && n.contains("蓝")));
    }
}
